GVFS/GVFS/CommandLine/HealthVerb.cs

```diff
@@ -0,0 +1,226 @@
+                    this.Directory = Environment.CurrentDirectory.Substring(enlistment.WorkingDirectoryRoot.Length);
+                }
+                else
+                {
+                    // If the path is not under the source root, set the directory to empty
+                    this.Directory = string.Empty;
+                }
+            }
+
+            this.Directory = this.Directory.Replace('\\', GVFSConstants.GitPathSeparator);
+
+            GVFSEnlistmentHealthCalculator.GVFSEnlistmentPathData pathData = new GVFSEnlistmentHealthCalculator.GVFSEnlistmentPathData();
+
+            this.GetPathsFromGitIndex(enlistment, pathData);
+            this.GetPlaceholdersFromDatabase(enlistment, pathData);
+            this.GetModifiedPathsFromPipe(enlistment, pathData);
+
+            pathData.NormalizeAllPaths();
+
+            GVFSEnlistmentHealthCalculator enlistmentHealthCalculator = new GVFSEnlistmentHealthCalculator(pathData);
+            GVFSEnlistmentHealthCalculator.GVFSEnlistmentHealthData enlistmentHealthData = enlistmentHealthCalculator.CalculateStatistics(this.Directory);
+
+            this.PrintOutput(enlistmentHealthData);
+        }
+
+        private void PrintOutput(GVFSEnlistmentHealthCalculator.GVFSEnlistmentHealthData enlistmentHealthData)
+        {
+            string trackedFilesCountFormatted = enlistmentHealthData.GitTrackedItemsCount.ToString("N0");
+            string placeholderCountFormatted = enlistmentHealthData.PlaceholderCount.ToString("N0");
+            string modifiedPathsCountFormatted = enlistmentHealthData.ModifiedPathsCount.ToString("N0");
+
+            // Calculate spacing for the numbers of total files
+            int longest = Math.Max(trackedFilesCountFormatted.Length, placeholderCountFormatted.Length);
+            longest = Math.Max(longest, modifiedPathsCountFormatted.Length);
+
+            // Sort the dictionary to find the most hydrated directories by percentage
+            List<KeyValuePair<string, decimal>> topLevelDirectoriesByHydration = enlistmentHealthData.DirectoryHydrationLevels.Take(this.DirectoryDisplayCount).ToList();
+
+            this.Output.WriteLine("\nHealth of directory: " + enlistmentHealthData.TargetDirectory);
+            this.Output.WriteLine("Total files in HEAD commit:           " + trackedFilesCountFormatted.PadLeft(longest) + " | 100%");
+            this.Output.WriteLine("Files managed by VFS for Git (fast):  " + placeholderCountFormatted.PadLeft(longest) + " | " + this.FormatPercent(enlistmentHealthData.PlaceholderPercentage));
+            this.Output.WriteLine("Files managed by git (slow):          " + modifiedPathsCountFormatted.PadLeft(longest) + " | " + this.FormatPercent(enlistmentHealthData.ModifiedPathsPercentage));
+
+            this.Output.WriteLine("\nTotal hydration percentage:           " + this.FormatPercent(enlistmentHealthData.PlaceholderPercentage + enlistmentHealthData.ModifiedPathsPercentage).PadLeft(longest + 7));

```

Comment:
Would it makes sense to move most of this code to the `GVFSEnlistmentHealthCalculator.GVFSEnlistmentHealthData` class with a `public string FormatForOutput(int directoriesToDisplay);`?

Reply:
I think it makes more sense to leave it in the verb since it is possible that this data will be consumed elsewhere and displayed or formatted differently, the verb in particular just displays it in this way. In general I like to leave the responsibilities for formatting and outputting to the client and only pass back data from the business logic.